op-node/rollup/derive/blob_data_source.go

```diff
@@ -145,13 +164,14 @@ func (ds *BlobDataSource) open(ctx context.Context) ([]blobOrCalldata, error) {
 // dataAndHashesFromTxs extracts calldata and datahashes from the input transactions and returns them. It
 // creates a placeholder blobOrCalldata element for each returned blob hash that must be populated
 // by fillBlobPointers after blob bodies are retrieved.
-func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address, logger log.Logger) ([]blobOrCalldata, []eth.IndexedBlobHash) {
+func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address, logger log.Logger, txSucceedMap map[common.Hash]bool) ([]blobOrCalldata, []eth.IndexedBlobHash) {
 	data := []blobOrCalldata{}
 	var hashes []eth.IndexedBlobHash
 	blobIndex := 0 // index of each blob in the block's blob sidecar
 	for _, tx := range txs {
 		// skip any non-batcher transactions or failed transactions
-		if !(isValidBatchTx(tx, config.l1Signer, config.batchInboxAddress, batcherAddr, logger)) {
+		// blobIndex needs to be incremented for both invalid batch tx and failed tx
+		if (!isValidBatchTx(tx, config.l1Signer, config.batchInboxAddress, batcherAddr, logger)) || (txSucceedMap != nil && !txSucceedMap[tx.Hash()]) {
```

Comment:
Consider a blob tx can be failed with successful blobs, I wonder if the following is a valid case, and we can handle it: 
the current tx is a valid batch tx but failed, while the blob it send is accessible from the beacon.

Reply:
It's exactly the case this PR ix fixing.

Reply:
If I understand right, the case being fixed is - the current tx is NOT a valid batch tx, and it failed.
The case I asked is - the current tx is a valid batch tx, and it failed. My question is: is this possible?

Reply:
It's possible, but there's no difference when dealing with `invalid batch tx/failed` or `valid batch tx/failed`.